### PR TITLE
[ROCKETMQ-243] BrokerData#selectBrokerAddr() picks the 1st element in the list of addresses

### PR DESCRIPTION
**Pull request by shroman**
https://issues.apache.org/jira/browse/ROCKETMQ-243

### CODE REVIEW

**Comment by Jaskey** [comment]
Broker contention should be safe but random pick can balance the network call for to all masters.

but we have to be cautions about this since for the current implementations, the broker addr will be very stable since it always return the same broker addr to the caller.

There are more than ten caller methods on this `selectBrokerAddr`, we have to make sure the random pick will not cause any side effect.

I believe it has not @lizhanhui @vongosling @zhouxinyu what's your opions.

@shroman , and since this method only returns the master addrs, I will suggest rename it to `pickMasterAddr`. Since the method is only called by some inner impl caller, I think it is safe to rename.

**Comment by shroman** [comment]
@Jaskey It returns a random broker address if the one with MASTER_ID is not found. Why do you suggest renaming?

I hope the execution flow doesn't depend on the current method implementation and broker address change is ok -- otherwise it is a huge flaw for a distributed system like this. Guys, please share your thoughts.

**Comment by Jaskey** [comment]
@shroman  Sorry I misunderstand this method that I though it returns the master address, since it is prefered to return the master address then slave, renaming seems to be not necessary while comment should be modified since it is only randomly pick while master is not found, thus only slave will be randomly pick.

**Comment by shroman** [comment]
@Jaskey I modified the comment, thanks!

File: common/src/main/java/org/apache/rocketmq/common/protocol/route/BrokerData.java [hunk 1]
```diff
@@ -37,15 +41,21 @@ public BrokerData(String cluster, String brokerName, HashMap<Long, String> broke
...
-        String value = this.brokerAddrs.get(MixAll.MASTER_ID);
-        if (null == value) {
-            for (Map.Entry<Long, String> entry : this.brokerAddrs.entrySet()) {
-                return entry.getValue();
-            }
+        String addr = this.brokerAddrs.get(MixAll.MASTER_ID);
+
+        if (addr == null) {
+            List<Long> keys = new ArrayList<Long>(brokerAddrs.keySet());
+            return brokerAddrs.get(keys.get(random.nextInt(keys.size())));
```
Comment (vsair) [inline common/src/main/java/org/apache/rocketmq/common/protocol/route/BrokerData.java]: why not use values to construct list directly?

Comment (shroman) [inline common/src/main/java/org/apache/rocketmq/common/protocol/route/BrokerData.java]: what exactly is the problem with the current approach?

Comment (Jaskey) [inline common/src/main/java/org/apache/rocketmq/common/protocol/route/BrokerData.java]: @shroman use `values()` will benefit from one less method call of `map.get(i)`

Comment (shroman) [inline common/src/main/java/org/apache/rocketmq/common/protocol/route/BrokerData.java]: Ok, changed to the list.

**Comment by dongeforever** [comment]
LGTM

**Comment by shroman** [comment]
Merged to develop branch.

**Pull request closed without merging**